src/main/java/cash/z/ecc/android/sdk/tool/SaplingParamTool.kt

```diff
@@ -0,0 +1,135 @@
+         * have write access to this directory. Typically, this should be the app's cache directory
+         * because it is not harmful if these files are cleared by the user since they are downloaded
+         * on-demand.
+         */
+        suspend fun fetchParams(destinationDir: String) = withContext(Dispatchers.IO) {
+            val client = createHttpClient()
+            var failureMessage = ""
+            arrayOf(
+                ZcashSdk.SPEND_PARAM_FILE_NAME,
+                ZcashSdk.OUTPUT_PARAM_FILE_NAME
+            ).forEach { paramFileName ->
+                val url = "${ZcashSdk.CLOUD_PARAM_DIR_URL}/$paramFileName"
+                val request = Request.Builder().url(url).build()
+                val response = client.newCall(request).execute()
+                if (response.isSuccessful) {
+                    twig("fetch succeeded")
+                    val file = File(destinationDir, paramFileName)
+                    if (file.parentFile.exists()) {
+                        twig("directory exists!")
+                    } else {
+                        twig("directory did not exist attempting to make it")
+                        file.parentFile.mkdirs()
+                    }
+                    Okio.buffer(Okio.sink(file)).use {
+                        twig("writing to $file")
+                        it.writeAll(response.body().source())
+                    }
+                    twig("fetch succeeded, done writing $paramFileName")
+                } else {
+                    failureMessage += "Error while fetching $paramFileName : $response\n"
+                    twig(failureMessage)
+                }
+            }
+            if (failureMessage.isNotEmpty()) throw TransactionEncoderException.FetchParamsException(
+                failureMessage
+            )
+        }
+
+
+        fun clear(destinationDir: String){
+            if(validate(destinationDir)){
+                arrayOf(
+                    ZcashSdk.SPEND_PARAM_FILE_NAME,
+                    ZcashSdk.OUTPUT_PARAM_FILE_NAME
+                ).forEach { paramFileName ->
+                    val file = File(destinationDir, paramFileName)
+                    if(file.deleteRecursively()){
+                        twig("Files deleted successfully")
+                    }else{
+                        twig("Error: Files not able to be deleted!")
+                    }
+                }
+            }
+        }
+
+        fun validate(destinationDir: String): Boolean{
+            arrayOf(
+                ZcashSdk.SPEND_PARAM_FILE_NAME,
+                ZcashSdk.OUTPUT_PARAM_FILE_NAME
+            ).forEach { paramFileName ->
+                val file = File(destinationDir, paramFileName)
+                if(file.exists()){
+                    twig("Files does  exist!")
+                    return true
+                }
+            }
+            twig("Error: File does not exist!")
+            return false
+        }
```

Comment:
It appears that this will return true if one file exists. We probably want to change this so that it only returns true when both files exist. The simplest way to do that would probably be to elevate the return with something like this:

```suggestion
    fun validate(destinationDir: String): Boolean {
        return arrayOf(
            ZcashSdk.SPEND_PARAM_FILE_NAME,
            ZcashSdk.OUTPUT_PARAM_FILE_NAME
        ).all { paramFileName ->
            File(destinationDir, paramFileName).exists()
        }.also {
            twig("Param files${if(!it) "did not"} both exist!")
        }
    }
```

Reply:
Can by any chance the user manually delete one of the files? What's the action to take in that case? if no action is being taken, quick fix could be 
```

if ! validate(path) {
  clear(path)
  fetchParams(path)
}
```

Reply:
> It appears that this will return true if one file exists. We probably want to change this so that it only returns true when both files exist. The simplest way to do that would probably be to elevate the return with something like this:

Committed your suggestions, made a few modifications because the code could not compile.

Reply:
sorry for the compile issue. I was typing the code on github but didn't test it. Looks like it just needed the `else` clause. Adding `else ""` would fix that:

```kotlin
.also {
    println("Param files${if(!it) "did not" else ""} both exist!")
}
```

The [code changes here](https://github.com/zcash/zcash-android-wallet-sdk/pull/206/commits/023d1fb079e9a719c765a1c1a549cbad4ed36ee1#diff-ade3efa42bf3ca434f2d51dedb711f797fd77ba76dd65d44968e0f52079878e3R107-R113) would result in two log lines, when a file is missing:

> Param files
> did not

Since we probably want one log line, instead of two for this, I would suggest either adding the `else ""` code from above or maybe simplifying the existing code to the following:

```kotlin
            }.also {
                if(it)
                    twig("Both param files exist!")
                else
                    twig("Both param files did not exist!")
            }
```

Reply:
Resolved by https://github.com/zcash/zcash-android-wallet-sdk/pull/206/commits/ccd1464bc0fc6b4f2bdc432e9eb224717a1f31f1